include/swift/Reflection/ReflectionContext.h

```diff
@@ -789,61 +789,150 @@ class ReflectionContext
+    return true;
+  }
+
+  bool getDynamicTypeAndAddressErrorExistential(
+      RemoteAddress object, const TypeRef **OutInstanceTR,
+      RemoteAddress *OutInstanceAddress, bool *IsBridgedError = nullptr) {
+    auto result = readMetadataAndValueErrorExistential(object);
+    if (!result)
+      return false;
+
+    auto typeResult =
+        readTypeFromMetadata(result->MetadataAddress.getAddressData());
+    if (!typeResult)
+      return false;
+
+    *OutInstanceTR = typeResult;
+    *OutInstanceAddress = result->PayloadAddress;
+    if (IsBridgedError)
+      *IsBridgedError = result->IsBridgedError;
+
+    return true;
+  }
+
+  bool
+  getDynamicTypeAndAddressOpaqueExistential(RemoteAddress object,
+                                            const TypeRef **OutInstanceTR,
+                                            RemoteAddress *OutInstanceAddress) {
+    auto result = readMetadataAndValueOpaqueExistential(object);
+    if (!result)
+      return false;
+
+    auto typeResult =
+        readTypeFromMetadata(result->MetadataAddress.getAddressData());
+    if (!typeResult)
+      return false;
+
+    *OutInstanceTR = typeResult;
+    *OutInstanceAddress = result->PayloadAddress;
+
+    return true;
+  }
+
   bool projectExistential(RemoteAddress ExistentialAddress,
                           const TypeRef *ExistentialTR,
                           const TypeRef **OutInstanceTR,
                           RemoteAddress *OutInstanceAddress,
                           remote::TypeInfoProvider *ExternalTypeInfo) {
-    if (ExistentialTR == nullptr)
+    auto ExistentialRecordTI =
+        getRecordTypeInfo(ExistentialTR, ExternalTypeInfo);
+    if (ExistentialRecordTI == nullptr)
       return false;
 
-    auto ExistentialTI = getTypeInfo(ExistentialTR, ExternalTypeInfo);
-    if (ExistentialTI == nullptr)
+    switch (ExistentialRecordTI->getRecordKind()) {
+    case RecordKind::ClassExistential:
+      return getDynamicTypeAndAddressClassExistential(
+          ExistentialAddress, OutInstanceTR, OutInstanceAddress);
+
+    case RecordKind::OpaqueExistential: {
+      return getDynamicTypeAndAddressOpaqueExistential(
+          ExistentialAddress, OutInstanceTR, OutInstanceAddress);
+    }
+    case RecordKind::ErrorExistential: {
+      return getDynamicTypeAndAddressErrorExistential(
+          ExistentialAddress, OutInstanceTR, OutInstanceAddress);
+    }
+    default:
       return false;
+    }
+  }
+
+  bool projectExistentialForLLDB(RemoteAddress ExistentialAddress,
+                                 const TypeRef *ExistentialTR,
+                                 const TypeRef **OutInstanceTR,
+                                 RemoteAddress *OutInstanceAddress) {
+    auto IsClass = [](const TypeRef *TypeResult) {
+      // When the existential wraps a class type, LLDB expects that the
+      // address returned is the class instance itself and not the address
+      // of the reference.
```

Comment:
Could this extra unwrapping be done in LLDB, or is it nested deep inside this function? Basically I wonder if we could get by with a single API here or if that isn't feasible.

Reply:
I don't see a clean way of doing this without altering the API of the original function, which is why I think keeping it as two separate functions might be the best option.